File: training-data/ruby/trie.rb
class TrieNode
  attr_accessor :children, :is_word

  def initialize
    @children = {}
    @is_word = false
  end
end

class Trie
  def initialize
    @root = TrieNode.new
  end

  def insert(word)
    node = @root
    word.each_char do |ch|
      node = (node.children[ch] ||= TrieNode.new)
    end
    node.is_word = true
  end

  def search(word)
    node = find(word)
    !node.nil? && node.is_word
  end

  def starts_with?(prefix)
    !find(prefix).nil?
  end

  private

  def find(s)
    node = @root
    s.each_char do |ch|
      node = node.children[ch]
      return nil if node.nil?
    end
    node
  end
end

trie = Trie.new
%w[cat car card care].each { |w| trie.insert(w) }
puts "#{trie.search('car')} #{trie.search('ca')} #{trie.starts_with?('ca')}"

File: training-data/cpp/trie.cpp
#include <iostream>
#include <memory>
#include <string>
#include <unordered_map>

struct TrieNode {
    std::unordered_map<char, std::unique_ptr<TrieNode>> children;
    bool isWord = false;
};

class Trie {
public:
    void insert(const std::string& word) {
        TrieNode* node = root.get();
        for (char ch : word) {
            if (!node->children.count(ch)) node->children[ch] = std::make_unique<TrieNode>();
            node = node->children[ch].get();
        }
        node->isWord = true;
    }

    bool search(const std::string& word) const {
        TrieNode* node = find(word);
        return node != nullptr && node->isWord;
    }

    bool startsWith(const std::string& prefix) const {
        return find(prefix) != nullptr;
    }

private:
    std::unique_ptr<TrieNode> root = std::make_unique<TrieNode>();

    TrieNode* find(const std::string& s) const {
        TrieNode* node = root.get();
        for (char ch : s) {
            auto it = node->children.find(ch);
            if (it == node->children.end()) return nullptr;
            node = it->second.get();
        }
        return node;
    }
};

int main() {
    Trie trie;
    for (const auto& word : {"cat", "car", "card", "care"}) trie.insert(word);
    std::cout << trie.search("car") << " " << trie.search("ca") << " " << trie.startsWith("ca") << std::endl;
    return 0;
}
